storage_controller/src/service.rs

```diff
@@ -1040,22 +1040,49 @@ impl Service {
 
         // In case a node was deleted while this reconcile is in flight, filter it out of the update we will
         // make to the tenant
-        result
-            .observed
-            .locations
-            .retain(|node_id, _loc| nodes.contains_key(node_id));
+        let deltas = result.observed_deltas.into_iter().flat_map(|delta| {
+            // In case a node was deleted while this reconcile is in flight, filter it out of the update we will
+            // make to the tenant
+            let node = nodes.get(delta.node_id())?;
+
+            if node.is_available() {
+                return Some(delta);
+            }
+
+            // In case a node became unavailable concurrently with the reconcile, observed
```

Comment:
What if it became unavailable, then available again, while a reconciler was 
running?  The way I'm reading it, the upsert we're passing through in the `return Some(...` above could still be stale.

To be totally robust, I think we would have to either send all observed state updates through the result channel, or keep some versioning on the .observed of a tenant shard.  Perhaps reconcile sequence would work for this?  The with that version, this piece of code could detect if updates were made concurrently with the reconcile, and in this case reset all the observed locations to None to reflect our uncertainty.

Reply:
> What if it became unavailable, then available again, while a reconciler was running?

I think it is still correct, but I agree it's not obvious.

If the node goes unavailable, then available again during a reconcile, then it means that all location config calls were successful. Hence, we should trust the observed state of the reconciler.

Reply:
I think this is robust enough for now, although I agree with you that it's hard to prove it's always correct.

Reply:
John and I discussed this on Friday a bit:
* This PR doesn't provide a generic solution because observed locations also carry generations
* We need a generic solution, but don't have to do it right now. This is an improvement and probably good enough.
* We can detect races when processing results by looking at generations

I've implemented the last bullet point in https://github.com/neondatabase/neon/pull/9276/commits/d522d88024224435e09095175df50fddf9a3e0ce and https://github.com/neondatabase/neon/pull/9276/commits/59af1d6cb751fc4c4fd7d38ec2151605fadf52b2. This should make it robust enough for now.